### DISCUSSION TITLE
Review of pipeline design

### DISCUSSION THREAD

**Discussion by jordansread** [USGS-R/ds-pipelines-targets-example-wqp#41] [General]
Great work- I took a careful look at this pipeline and I like what I see. I do have some questions and suggestions that may either make some parts of this simpler or more robust. I think there are probably a number of design decisions that were made that make sense, but I may be missing the context for so some of these may be known to the group already:

#### 1) Can you drop the CONUS and `tigris::states` dependency by using `offset` in the grid?
One limitation of the current pipeline is the reliance on a CONUS boundary to set up the cells for inventory. I think you may want to consider modifying this for two reasons: 1) the [CONUS boundaries](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/2e8487f370c2e16252b58f7a23b49d90f2249540/1_fetch/src/create_grids.R#L24) will exclude use of the pipeline in AK, HI, PR and other places where WQP data exists, such as Guam, 2) the edge of the CONUS map is a little arbitrary and results in cells that aren't referenced from an integer (which can cause another odd issue as documented [here](https://github.com/USGS-R/dataRetrieval/issues/616) ) and the grid cells look like this: 
```r
sf::st_bbox(p1_conus_grid[1,1])
      xmin       ymin       xmax       ymax 
-124.84897   24.39631 -123.84897   25.39631 
```
and small updates or different years of the `tigris::states` `year` could nudge this in a small way and invalidate the inventory even if the change isn't helpful to the WQP inventory itself. Instead, using something like `sf::st_make_grid(p1_AOI_sf, offset = c(-180, -90), crs=4326, cellsize = c(1,1))` to make your grid can be more inclusive and rebuilds of inventory on cells won't be as greedy. Using an `offset` will result in the grid starting at the same point for all builds and then filtering to the overlap with AOI. I've found that approach to give the user more control over when they want to have the grids change vs stay fixed but perhaps add a new active target branch. 

#### 2) Do you need to split Char groups for inventory or data pulls? the main groups (e.g., `param_groups_select`)
This is a design question related to the use of many variables. I'm not sure what the right answer is, but if I understand the pipeline correctly, you are including all parameters in the config and then using certain ones for the pull (with `param_groups_select`) but the inventory and the data pull itself pulls all variables at once, even if they aren't as related. An alternative is to branch off the major categories, so your [current example](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/2e8487f370c2e16252b58f7a23b49d90f2249540/_targets.R#L15) would have inventories for each cell-group combination instead of just each cell. Again, a design decision that has benefit in the way you've done it for simplicity, but the tradeoff as I see it is if I change `param_groups_select <- c('temperature','conductivity')` to `param_groups_select <- c('temperature','conductivity','nitrogen')` it will render most of the pipeline stale instead of just adding a few new inventory and data pull targets that pertain to the new char group. 

#### 3) There are two sections of buffering and perhaps you could remove them or make the functionality less visible
I like the flexibility exposed for power users when you can specify a buffer for [cells](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/2e8487f370c2e16252b58f7a23b49d90f2249540/1_fetch.R#L65) and [sites/AOI](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/main/1_fetch.R#L88) but it does expose some complexity in spatial filtering (e.g., using no buffer, you are relying on the WQP service to do the spatial filtering) and also some edge cases that could be confusing, such as the situation where someone chose a larger `buffer_dist_m` than their `dist_m`, which could result in lost sites that would be in the buffer to the AOI but otherwise were left out of the cell query. My suggestion would be to give both of these function a default of `0` for the buffer and don't use the arguments in the example itself, but keep the functionality flexible to include it. 

#### 4) p1_wqp_data_aoi from `fetch_wqp_data()` is not impacted by resultCounts changing in edge-cases
The inputs to `fetch_wqp_data()` are a data.frame which has `site_id, site_n, download_grp, tar_group`. This made me slightly concerned about the situation where the upstream inventory result changed but the resulting data.frame here remained the same, therefore no new data were pulled. Including `resultCount` as another column would clear this up I think. 

To illustrate - if we run a full build and then two weeks later run another with a tiny change to the AOI that returns the same number of sites in a given cell, but site X now has 50 more samples. If the order of sites is still the same and the site numbers are still the same, this data.frame is going to look the same to `targets` and be "fresh" and we won't get the new 50 samples. These situations are rare with this particular pipeline, but designing the pipeline in a way where the rebuilds are understood and predictable can help us keep good patterns in other situations where they'd be more confusing and common (e.g., this is why we use hash tables in early gen pipelines because they pair the vector file name with the file's hash value, because just tracking the vector of names can hide changes in the underlying file). 

#### 5) How does the pipeline handle single sites with huge amounts of data?
I didn't test what happens to the pipeline if we include a site with a massive amount of data in the pull. In a [different WQP pipeline](https://github.com/USGS-R/lake-temperature-model-prep/blob/1ff712ec058fe57d978adbba0861e8e2fcb7f13a/6_wqp_fetch/src/fetch_wqp_data.R#L27), I use two ways to threshold the sites that will be in a given pull - the total results expected _and_ the maximum number of sites we'd want in a given pull. In the current pipeline, the sites are lumped according to max sites per pull. If you have one of the 500 sites with a **ton** of data, what happens in this case? Does this target fail a lot? I'm not really sure with the current WQP services and therefore not sure this situation is more robust now. 

You could consider lumping sites for data pulls according to cell, resultCount, and max sites...but that is again another complexity design decision. One reason I'd advocate for doing this is that it makes it easier to expand spatially without invalidating the earlier pull (e.g., adding a cell to the inventory only adds that cell's worth of sites to the inventory and that cell's inventory data to the data pull instead of wiping out the inventory/pull completely). But, there is an argument to be made that the pulls should be complete each time they are triggered so you can communicate which day the WQP was accessed instead of having a different answer to this depending on each cell...

Minor feedback

- consider making `dist_m = 0` a default arg [here](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/2e8487f370c2e16252b58f7a23b49d90f2249540/1_fetch/src/create_grids.R#L56) since I'd suggest the default pipeline doesn't have buffering and this arg can be left out of the target
- in [this message](https://github.com/USGS-R/ds-pipelines-targets-example-wqp/blob/2e8487f370c2e16252b58f7a23b49d90f2249540/1_fetch/src/get_wqp_inventory.R#L30) perhaps change to `'Inventorying WQP data for grid %s'` to be more consistent with the target name language?
- can `p1_wqp_params_yml` and `p1_wqp_params` be just one target? Perhaps just using `p1_wqp_params`? Perhaps it needs to be two because the file type needs to be declared to `targets`.

**Comment by lekoenig** [comment]
Thanks for this great feedback, @jread-usgs! 

I've created a few issues based on these questions and suggestions:

- [x] point 1 above addressed in issue #42 
- [x] point 3 above addressed in issue #43 
- [x] point 4 above addressed in issue #44

I'll give points 2 and 5 a bit more thought in terms of how we might tackle those, perhaps as part of #36.

**Comment by lekoenig** [comment]
> can p1_wqp_params_yml and p1_wqp_params be just one target? Perhaps just using p1_wqp_params? Perhaps it needs to be two because the file type needs to be declared to targets.

We could combine them, but my instinct is that we would need to retain `p1_wqp_parms_yml` as a file target in order for {targets} to track any changes to that file. @lindsayplatt, can you confirm whether that's the case?

**Reply by lindsayplatt** [reply]
Right - any "in" file needs to first be declared as a file target so that the file is tracked by targets. Then, you can read in and use it. If you want only one target, the alternative could be to skip `p1_wqp_params` and just load the yml when it is used inside other functions. I think these are fairly inexpensive targets to build, so my suggestion would be to keep them as-is.

@jread-usgs to use some `scipiper` lingo, excluding the first target `p1_wqp_params_yml` and just doing the second would be equivalent to writing `I('1_fetch/cfg/wqp_codes.yml')`, so the file changes would be ignored.
